### ISSUE TITLE
[Feature Request] Configurable buttons mappings via Bluetooth

### ISSUE THREAD

**Issue by ste2425** [ricardoquesada/bluepad32#30]
It would be great if there was a way to connect with the Arduino via web bluetooth and issue commands to create custom button mappings.

This would allow consumers of this library to build web apps that allow users of their projects to create custom button mappings.

BlueRetro solve the configuration problem this way. You put the board into configuration mode. This means controllers cannot be connected but you can then connect a web app they provide via web bluetooth. Within this app you can then create mappings.

For this request i don't think we need the web app, just the means to connect to the Arduino. 

I'm a nodejs dev so if/when this feature request is implemented i can build a Node based API that can run in node and the browser to communicate with Bluepad if that is helpful.

**Comment by ste2425** [comment]
Technically for my specific needs i dont need Bluepad to handle the configuration.

Iv'e built a PSP consolizer and manually handle toggleable buttons mappings between the controller and the PSP. So personally all I'd need is a way of sending messages with a payload between a Web Bluetooth powered web app, through Bluepad32 and into my Arduino code. I could then handle the mappings myself.

Not too sure of the work involved in this feature but that may be a possible alternative, or first pass rather than implementing the entire custom configuration system within BluePad if thats allot of work to implement.

**Comment by ricardoquesada** [comment]
Let's discuss here since we already have most of the context

**Comment by ste2425** [comment]
Oki Doki 

So lets start with what i have.

A Arduino communicating with a PSP. I hold a series of mappings between controller buttons returned from BluePad to the corresponding PSP buttons to trigger. I have three mappings a can toggle through a controller button combo to best suite different games. Platformers, FPS and PS1 games.

I wish to build a feature where users of my consolizer can build their own button mappings to better suite specific games. I will extend it in the future to have the PSP and Arduino communicate over serial so the Arduino knows which game is being played and automatically select a mapping which has been bound to that game ID.

So i planned to build an app (either browsers or electron based) that allowed users to build these custom controller to PSP mappings then send them to my Arduino.

Issue is how to communicate with the Arduino. I could use WIFI, but Bluepad hasn't got that implemented. I could use BLE (as weve been discussing) or i could use Serial.

Out of them all i this Serial offers the worst user experience. Especially when my final mod will fit inside the PSP so plugging a USB cable into the Arduino will be difficult.

So you can see i think i may some control over the mappings and instruct BluePad of the mappings but all the wireless means of communicating are unavailable to me within my Arduino land.

**Comment by ricardoquesada** [comment]
I think I got it... and IIUC I think we have something with a similar end result using another way...

...I'll post the details about it tonight

**Comment by ricardoquesada** [comment]
TL;DR: Let's store the database on gamepads in the ESP32, and add API to apply/delete mapping from Arduino.

IIUC:

Users are going to create mappings and send it to Arduino.

Let's assume that somehow Arduino has all the mappings now.
How Arduino is going to tell the ESP32 to use the mapping?
A possible solution would be to add a function called:

`BP32.applyGamepadMappings(gamepad_idx, struct gamepad_mappings)`

And it goes over serial.... and that is valid.

But another approach is that the ESP32 already has a collection of mappings. Let's say there is a dictionary like this:
```// pseudo code
  // Data stored in NVS in the ESP32
 map {
     "metal_gear":   [struct with mapping],
     "grand theft auto": [struct with mappings],
 }

// From Arduino you will call:
BP32.applyGamepadMappings(gamepad_idx 1, "metal_gear");
```
How do the gamepad mappings gets stored in the first place ?

One possible way, is that the BLE service allows you to upload a gamepad mapping, and you give it a name.
And it gets stored there in NVS.

But how to notify Arduino that a new mapping was added?

We can add a few functions, perhaps something like:
`BP32.listGamepadMappings() -> returns list[names]` where name is the name of the mapping


Would that work?

**Comment by ste2425** [comment]
That would work, depending how how configurable the mappings are (i havent seen what they can or can't do so going at that a little blind).

For my specific use case i can map controller buttons to PSP buttons. Controller analog sticks (either if there are multiple) to the PSP's analog stick. However i can also map controller buttons to the PSP analog stick and can map controller analog sticks (either) to the PSP buttons.

If the mappings system within Bluepad allows for that level of mappings thats perfect, if were talking about how we map different physcial controllers (DS4, Xbox etc) to the standard BluePad controller layout that may not work.

The other requirement is i have extra config i would need to set within the PSP consolizer webapp and send to my Arduino via BLE which is not really a BluePad concern (i don't think). For example i have a configurable deadzone on the analog sticks. When an analog stick is mapped to a PSP button how far the stick has to be moved to trigger the button is configurable. 

I also have other config not related to the controller such as auto booting the PSP when a controller is connected, auto enabling TVOut on the PSP things like that.

I think my only concern is the comms between the web app and BluePad will stop there then we could build a system to manage the mappings config, but not that extra PSP domain specific config.

Maybe if the BLE comms has to stop at BLuePad and my Arduino code has to talk to BluePad directly we could also have some form meta data? A basic key value dictonary? That way my extra specific to me config could be set within that meta and ready from BluePad via a meta data specific method?

However what you described there looks like it will really help with the mappings if we can figure out the extra me specific config

Thanks for working with me on this, trying to suggest solutions that are generic but also solve my specific problem :)

**Comment by SpudGunMan** [comment]
Sounds like a cool patch! @ricardoquesada

**Comment by ricardoquesada** [comment]
> However i can also map controller buttons to the PSP analog stick and can map controller analog sticks (either) to the PSP buttons

This is currently not supported, but I guess it can be done more or less easy

> The other requirement is i have extra config i would need to set within the PSP consolizer webapp and send to my Arduino which is not really a BluePad concern (i don't think)

This could also be part of Bluepad32... but might take a while... if you need it in the short term, better if you implement it in the Arduino side... but we will still have the problem of how to send that data from ESP32 to Arduino.

Perhaps adding a "raw" BLE service on the ESP32 might work.
By that I mean that whatever is received on ESP32, the ESP32 will send it to the Arduino... the data will be opaque for the ESP32... and the consumer (in this case Arduino) is reponsible for parsing it correctly.

And you can use it for whatever you want... mappings, extra config, etc...
this way I won't block you... you can start implementing the mappings using this "raw" service... and once the mappings are implemented in BP32 you can use them ... or not.. and keep using your implementation.

The way I see it is something like this:

- From your client (node.js) you send 64 bytes of data to the BLE "raw data" service running on ESP32.
- ESP32 will store it in volatile memory... it will be  erased in the next reboot
- From your Arduino core you query with something like `BP32.fetchRawData()` and will return a blob of max 64 bytes (or empty)
- And if you want to send more data, you just repeat the process

Would that work?

**Comment by ste2425** [comment]
That sounds perfect, the `raw` BLE service. Gives the best of both worlds.

That process sounds spot on, much better idea than what i was suggesting.

Next question, anything i can do to help!

**Comment by ricardoquesada** [comment]
If you want to implement it yourself, go ahead :) (you have to be familiar with Bluepad32 + BtStack code)....  otherwise I'll implement myself. And just keep an eye on this ticket... I'll post my progress here

**Comment by ste2425** [comment]
Nope not familiar at all, its all yours! Thanks :)

**Comment by ste2425** [comment]
Don't want to presume this feature request was only for me and only of use for me. However if it is i think its worth mentioning i don't believe i need it anymore.

I have migrated to the ESP32 platform so i can access to the ESP's flash storage and once #37 is complete will have examples of communicating with the ESP via BLE.

If preferred this issue can be closed.

**Comment by ricardoquesada** [comment]
sounds good. Let's close it.
